### DISCUSSION TITLE
What should I do if I need an overridable struct to store data?

### DISCUSSION THREAD

**Discussion by TreapGoGo** [Cyfrin/foundry-full-course-cu#730] [Q&A] [answered]
Sometimes we may meet such a condition in development:

A sub contract which inherits from a parent contract needs to override some of the functions. However, the type of the return value of the sub contract inherits from the type of the return value of the parent contract, on the basis of which it may add some new attributes (or state variables). The return value only needs to store some complex data and it doesn't need to have any methods or functions.

As we know, structs and libraries cannot override in Solidity. But if we use `contract` as the return value (of course it can inherit and override and add some new attributes), we have to deploy an instance of the return value contract on chain every time we `new` it. That is a waste of gas! Because we just need to store some data instead of deploying a big contract.

So how to deal with this problem? Is there a gas efficient way to implement complex overridable data types?

**Comment by 0xTagger** [comment, marked as answer]
You can use a combination of contract inheritance and storage pointers. Instead of deploying a new instance of the return value contract every time you need to create a new one - you can use a pattern where the sub contract references the storage of the parent contract. Another thing to keep in mind is by using immutable and constant state variables. These can help optimize gas usage because immutable variables are only stored once in the contract bytecode and constant variables do not require storage space at all.

Another possible method is to use an assembly language such as Yul which can be written inside Solidity contracts. Yul provides a way to directly access the EVM which allows for potentially more gas-efficient code. Please be aware that using an assembly language can be risky because it bypasses Solidity’s safety checks and can result in difficult to find bugs and so it should only be used when necessary.

**Reply by TreapGoGo** [reply]
Thanks so much for your answer!
